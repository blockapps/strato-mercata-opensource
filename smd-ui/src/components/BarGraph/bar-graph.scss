@import "../../../node_modules/@blueprintjs/core/dist/variables.scss";

.grid-lines {
  color: '#eee' !important;
  border-color: '#eee';
  background-color: '#ddd';
}
