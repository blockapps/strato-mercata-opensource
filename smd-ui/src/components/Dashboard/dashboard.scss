@import "../../../node_modules/@blueprintjs/core/dist/variables.scss";

.network-status {
  color: $pt-intent-success;
  margin-top: 0.7em;
  font-size: 20px;
  font-weight: bold;
  background-color: $pt-dark-code-background-color;
  padding: 5px 10px;
  text-align: center;
}
.full-width {
  display: "inline-block";
  width: 100%;
}

.bad-metric {
  color: #f74b4b;
}

.good-metric {
  color: #f5f8fa;
}

.metric {
  display: inline-block;
}

.healthy-network {
  color: #62d96b;
}

.unhealthy-network {
  color: #ffc940;
}

#networkInfo {
  display: flex;
  flex-direction: column;
}
