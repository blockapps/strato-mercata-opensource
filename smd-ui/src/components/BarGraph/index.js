import React, {Component} from 'react';
import * as Plottable from 'plottable';
import './bar-graph.css';

class BarGraph extends Component {

  constructor(props) {
    super(props);
    this.state = {
      plot: null,
      dataset: null,
      interaction: null,
      tooltipAnchor: null,
      tooltipText: null,
      tooltipBox: null,
      tooltip: null,
      firstRender: true
    }
  }

  maxY() {
    var max = 0;
    var y = 0;
    for (; y < this.props.data.length; y++) {
      max = Math.max(this.props.data[y].y, max);
    }
    return max;
  }

  minY() {
    var min = 0;
    var y = 0;
    for (; y < this.props.data.length; y++) {
      min = Math.min(this.props.data[y].y, min);
    }
    return min;
  }

  averageY() {
    var y = 0;
    this.props.data.forEach(val => {
      y += val.y;
    })
    return Math.round(y / this.props.data.length); //rounded or keep decimal?
  }

  renderGraph() {
    let min = this.minY();
    let max = this.maxY();
    let scaleMax = max + ((max - min) / 50 + 1);
    let scaleMin = min - ((max - min) / 50 + 1) < 0 ? 0 : min - ((max - min) / 50 + 1);
    const xScale = new Plottable.Scales.Linear().domain([0, 16]);
    const yScale = new Plottable.Scales.Linear().domain([scaleMin, scaleMax]);
    let self = this;

    // TODO: fix this. Do not mutate state directly. Use redux.
    // eslint-disable-next-line
    this.state.dataset = new Plottable.Dataset(this.props.data);
    // eslint-disable-next-line
    this.state.plot = new Plottable.Plots.Bar()
      .addDataset(this.state.dataset)
      .x(function (d) {
        return d.x;
      }, xScale)
      .y(function (d) {
        return d.y;
      }, yScale)
      //.animator(Plottable.Plots.Animator.MAIN, new Plottable.Animators.Easing().easingMode('quad'))
      .animated(true)
      .attr('fill', '#00AEEF')
      ;

    // eslint-disable-next-line
    this.state.interaction = new Plottable.Interactions.Pointer();
    this.state.interaction.onPointerMove(function(p){
      const entity = self.state.plot.entityNearest(p);
      if(entity) {
        self.state.tooltipText
          .text(entity.datum.y);
        const bbox = self.state.tooltipText.node().getBBox();
        self.state.tooltipBox
          .attr('x',bbox.x-8)
          .attr('y',bbox.y-3)
          .attr('width',bbox.width+16)
          .attr('height',bbox.height+6);
        self.state.tooltip
          .attr('points', [
            0,
            -20,
            -5,
            -15,
            5,
            -15
          ].join(','));

        self.state.tooltipAnchor
          .attr('transform','translate(' + entity.position.x + ',' + (entity.position.y+20)  + ')')
          .attr('opacity',1);
      }
    });

    this.state.interaction.onPointerExit(function(){
      self.state.tooltipAnchor
        .attr('opacity',0);
    });

    this.state.interaction.attachTo(this.state.plot);
    this.state.plot.renderTo("div#bg" + this.props.identifier);

    // eslint-disable-next-line
    this.state.tooltipAnchor = this.state.plot
      .foreground()
      .style('overflow','visible')
      .append('g')
      .attr('opacity', 0);

    // eslint-disable-next-line
    this.state.tooltipBox = this.state.tooltipAnchor
      .append('rect')
      .style('fill','#293742')
      .attr('rx', 4)
      .attr('ry', 4);

    // eslint-disable-next-line
    this.state.tooltip = this.state.tooltipAnchor
      .append('polygon')
      .style('fill','#293742');

    // eslint-disable-next-line
    this.state.tooltipText = this.state.tooltipAnchor
      .append("text")
      .style('fill','#f5f8fa')
      .attr('text-anchor','middle')
      .text('test');

  }

  componentDidMount() {
    this.renderGraph();
  }

  componentDidUpdate() {
    let min = this.minY();
    let max = this.maxY();
    let scaleMax = max + ((max - min) / 50 + 1);
    let scaleMin = min - ((max - min) / 50 + 1) < 0 ? 0 : min - ((max - min) / 50 + 1);
    const xScale = new Plottable.Scales.Linear().domain([0, 16]);
    const yScale = new Plottable.Scales.Linear().domain([scaleMin, scaleMax]);

    this.state.plot
      .x(function (d) {
        return d.x;
      }, xScale)
      .y(function (d) {
        return d.y;
      }, yScale)
      .animated(this.state.firstRender);
    this.state.dataset
      .data(this.props.data);
    // eslint-disable-next-line
    this.state.firstRender = false;
  }

  render() {

    return (
      <div className="pt-card pt-dark">
        <div className="row">
          <div className="col-sm-12 text-center">
            <h4>{this.props.label}</h4>
          </div>
        </div>
        <div className="row">
          <div className="col-sm-12">
            <div id={"bg" + this.props.identifier} style={{height: '120px'}}></div>
          </div>
        </div>
        <div className="row">
          <div className="col-sm-12 text-center">
            <h2>Avg: {' '}
                {this.averageY() !== this.averageY() ? "No Blocks" : this.averageY() +
                (this.props.units === undefined ? '' : ' ' + this.props.units)
                }
            </h2>
          </div>
        </div>

      </div>
    );
  }
}

export default BarGraph;
