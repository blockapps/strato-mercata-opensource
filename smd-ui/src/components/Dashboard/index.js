import React, { Component } from "react";
import { connect } from "react-redux";
import { withRouter, Link } from "react-router-dom";
import NodeCard from "../NodeCard";
import TransactionList from "../TransactionList";
import NumberCard from "../NumberCard";
import mixpanelWrapper from "../../lib/mixpanelWrapper";
import { endTour } from "../Tour/tour.actions";
// import { callAfterTour } from '../Tour/tour.helpers';
// import Tour from '../Tour';
import Tour from "../Tour";
import { env } from "../../env";
import io from "socket.io-client";
import BarGraph from "../BarGraph";
import PieChart from "../PieChart";
import "./dashboard.css";
import { hideLoading } from "react-redux-loading-bar";
import { subscribeRoom, unSubscribeRoom } from "../../sockets/socket.actions";
import { changeHealthStatus } from "./dashboard.action";
import {
  LAST_BLOCK_NUMBER,
  USERS_COUNT,
  CONTRACTS_COUNT,
  TRANSACTIONS_COUNT,
  BLOCKS_PROPAGATION,
  BLOCKS_DIFFICULTY,
  BLOCKS_FREQUENCY,
  TRANSACTIONS_TYPE,
  GET_NODE_UPTIME,
  GET_HEALTH,
  GET_SYSTEM_INFO,
  GET_SHARD_COUNT,
  GET_NETWORK_HEALTH,
} from "../../sockets/rooms";
import { sec2Date } from "../../lib/formatSeconds";
import ReactGA from "react-ga4";
import { Popover, PopoverInteractionKind, Position } from "@blueprintjs/core";
import ValidatorsCard from "../ValidatorsCard";

const socket = io(env.SOCKET_SERVER, {
  path: "/apex-ws",
  transports: ["websocket"],
});
// TODO: these should be part of a reducer state. Do the same for other global variables.
const tourSteps = [
  {
    title: "Welcome to STRATO!",
    text: "<strong>STRATO</strong> makes it easy to create and manage your custom blockchains.<br><br><strong>Ready to get started?</strong>",
    selector: "#tour-welcome",
    position: "bottom",
    type: "hover",
    isFixed: true,
  },
  {
    title: "Adding Users",
    text: "Before you write a Smart Contract, you must add some users that your Smart Contract will interact with.<br><br>For example, if you are splitting the bill for a monthly apartment rental, you might add <i>Roommate 1</i>, <i>Roommate 2</i>, and <i> Roommate 3.</i>",
    selector: "#accounts",
    position: "bottom",
    type: "hover",
    isFixed: true,
  },
];

class Dashboard extends Component {
  constructor(props) {
    super(props);
    this.state = {
      isHovering: false,
    };
  }
  componentDidMount() {
    this.props.subscribeRoom(LAST_BLOCK_NUMBER);
    this.props.subscribeRoom(USERS_COUNT);
    this.props.subscribeRoom(CONTRACTS_COUNT);
    this.props.subscribeRoom(BLOCKS_PROPAGATION);
    this.props.subscribeRoom(BLOCKS_FREQUENCY);
    this.props.subscribeRoom(BLOCKS_DIFFICULTY);
    this.props.subscribeRoom(TRANSACTIONS_COUNT);
    this.props.subscribeRoom(TRANSACTIONS_TYPE);
    this.props.subscribeRoom(GET_SHARD_COUNT);
    this.props.subscribeRoom(GET_SYSTEM_INFO);
    this.props.subscribeRoom(GET_NETWORK_HEALTH);

    mixpanelWrapper.track("dashboard_page_load");
    ReactGA.send({
      hitType: "pageview",
      page: "/dashboard",
      title: "Dashboard",
    });
  }

  displaySystemMetrics(cpu, memory, filesystem, networkStats, systemHealth) {
    if (cpu && memory && filesystem && networkStats) {
      //fileSystem data is sorted in descending order, max is first
      const maxUsageFilesystem = filesystem[0];

      return (
        <div
          className="sys-info-container col-sm-6 text-right row"
          style={{
            marginTop: "25px",
            marginBottom: "10px",
            marginLeft: "0px",
            marginRight: "0px",
          }}
        >
          <p className="text-right">
            <span
              id="cpuCurrentMetric"
              className={`metric ${
                cpu.currentLoad.isHealthy ? "good-metric" : "bad-metric"
              }`}
            >
              {`CPU: ${cpu.currentLoad.value.toFixed(2)}%`}
            </span>
            {" | "}
            <span
              id="cpuAvgMetric"
              className={`metric ${
                cpu.avgLoad.isHealthy ? "good-metric" : "bad-metric"
              }`}
            >
              {`Avg CPU : ${cpu.avgLoad.value.toFixed(2)}`}
            </span>
            {" | "}
            <span
              id="memoryMetric"
              className={`metric ${
                memory.use.isHealthy ? "good-metric" : "bad-metric"
              }`}
            >
              {`Mem: ${memory.use.value.toFixed(2)}%`}
            </span>
            {" | "}
            <span
              id="diskpaceMetric"
              className={`metric ${
                maxUsageFilesystem.use.isHealthy ? "good-metric" : "bad-metric"
              }`}
            >
              {`Disk: ${maxUsageFilesystem.use.value.toFixed(2)}%`}
            </span>
            {" | "}
            {systemHealth ? (
              <i className="fa-solid fa-circle-check healthy-network" />
            ) : (
              <i className="fa-solid fa-circle-exclamation unhealthy-network" />
            )}
          </p>
        </div>
      );
    } else {
      return (
        <div
          className="sys-info-container row col-sm-6 text-right"
          style={{
            marginTop: "25px",
            marginBottom: "10px",
            marginLeft: "0px",
            marginRight: "0px",
          }}
        >
          <p className="text-right">System Metrics Loading...</p>
        </div>
      );
    }
  }

  componentWillUnmount() {
    this.props.unSubscribeRoom(LAST_BLOCK_NUMBER);
    this.props.unSubscribeRoom(USERS_COUNT);
    this.props.unSubscribeRoom(CONTRACTS_COUNT);
    this.props.unSubscribeRoom(BLOCKS_PROPAGATION);
    this.props.unSubscribeRoom(BLOCKS_FREQUENCY);
    this.props.unSubscribeRoom(BLOCKS_DIFFICULTY);
    this.props.unSubscribeRoom(TRANSACTIONS_COUNT);
    this.props.unSubscribeRoom(TRANSACTIONS_TYPE);
    this.props.unSubscribeRoom(GET_HEALTH);
    this.props.unSubscribeRoom(GET_NODE_UPTIME);
    this.props.unSubscribeRoom(GET_SYSTEM_INFO);
    this.props.unSubscribeRoom(GET_SHARD_COUNT);
    this.props.unSubscribeRoom(GET_NETWORK_HEALTH);
  }

  render() {
    const difficultyData = this.props.dashboard.blockDifficulty;
    const txCount = this.props.dashboard.transactionsCount;
    const blockPropData = this.props.dashboard.blockPropagation;
    const txTypeData = this.props.dashboard.transactionTypes;
    const { usersCount, contractsCount, lastBlockNumber } =
      this.props.dashboard;
    const uptime = this.props.dashboard.uptime;
    const health = this.props.dashboard.health;
    const healthStatus = this.props.dashboard.healthStatus;
    const healthIssues = this.props.dashboard.healthIssues;
    const systemHealth = this.props.dashboard.systemStatus;
    const systemWarnings = this.props.dashboard.systemWarnings;
    const { cpu, memory, filesystem, networkStats } =
      this.props.dashboard.systemInfo || {};
    const synced = this.props.appMetadata.metadata
      ? this.props.appMetadata.metadata.isSynced
      : false;
    const metadata = this.props.appMetadata.metadata;
    const networkHealth = this.props.dashboard.networkStatus;
    const networkStatusMessage = this.props.dashboard.networkStatusMessage;

    return (
      <div className="container-fluid pt-dark" id="tour-welcome">
        <Tour
          name="dashboard"
          finalStepSelector="#accounts"
          nextPage="accounts"
          steps={tourSteps}
        />
        <div className="row d-flex align-items-center">
          <div className="col-sm-6 text-left">
            <h3>Node Stats</h3>
          </div>
          {this.displaySystemMetrics(
            cpu,
            memory,
            filesystem,
            networkStats,
            systemHealth
          )}
        </div>
        <div className="row">
          <div className="col-sm-12">
            <br />
          </div>
        </div>
        <div className="row">
          <div className="col-sm-4">
            <Popover
              isDisabled={synced && health && systemHealth}
              interactionKind={PopoverInteractionKind.HOVER}
              position={Position.BOTTOM}
              className={"full-width"}
              content={
                <div
                  className={`pt-dark pt-callout smd-pad-8 pt-icon-info-sign pt-intent-${
                    !metadata
                      ? "danger"
                      : !health || !systemHealth || !synced
                      ? "warning"
                      : "success"
                  }`}
                >
                  <h5 className="pt-callout-title">
                    {!metadata ? "API Disconnected" : healthStatus}
                  </h5>
                  {!metadata
                    ? "Cannot connect to the Node's API"
                    : !health || !systemHealth
                    ? `Health issues: ${
                        healthIssues.length > 0
                          ? healthIssues.join(". ")
                          : "unknown issue."
                      }`
                    : "Connected to STRATO Mercata"}
                </div>
              }
            >
              <NumberCard
                number={!metadata ? "DISCONNECTED" : healthStatus}
                description={sec2Date(uptime)}
                mode={
                  !metadata
                    ? "danger"
                    : !health || !systemHealth || !synced
                    ? "warning"
                    : "success"
                }
                iconClass={
                  !metadata
                    ? "fa-triangle-exclamation"
                    : !health || !systemHealth
                    ? "fa-exclamation-circle"
                    : !synced
                    ? "fa-rotate"
                    : "fa-check-circle"
                }
              />
            </Popover>
          </div>
          <div className="col-sm-4">
            <Link to="/accounts">
              <NumberCard
                number={usersCount}
                description="Users"
                iconClass="fa-users"
                className="smd-pointer"
              />
            </Link>
          </div>

          <div className="col-sm-4">
            <Link to="/contracts">
              <NumberCard
                number={contractsCount}
                description="Contracts"
                iconClass="fa-file-contract"
                className="smd-pointer"
              />
            </Link>
          </div>
        </div>
        <div className="row">
          <div className="col-sm-12">
            <br />
          </div>
        </div>
        <div className="row">
          <div className="col-sm-4">
            <NumberCard
              number={
                this.props.appMetadata && this.props.appMetadata.nodeInfo ? (
                  <div>
                    <p>
                      {this.props.appMetadata.nodeInfo.organization}{" "}
                      {this.props.appMetadata.nodeInfo.organizationalUnit}
                    </p>
                    <p>{this.props.appMetadata.nodeInfo.commonName}</p>
                  </div>
                ) : (
                  "No Identity"
                )
              }
              description="Node ID"
              iconClass={
                this.props.appMetadata && this.props.appMetadata.nodeInfo
                  ? "fa-id-card"
                  : "fa-exclamation-circle"
              }
              className={`smd-pointer`}
              mode={
                this.props.appMetadata && this.props.appMetadata.nodeInfo
                  ? ""
                  : "pt-intent-warning"
              }
              textSize="h4"
            />
          </div>
          <div className="col-sm-8">
            <NodeCard />
          </div>
        </div>

        <div className="row">
          <div className="col-sm-12">
            <hr />
          </div>
        </div>
        <div className="row">
          <div className="col-sm-9 text-left">
            <h3>Network Stats</h3>
          </div>
        </div>
        <div className="row">
          <div className="col-sm-3">
            <div id="networkInfo">
              <Link to="">
                <NumberCard
                  number={networkStatusMessage || "UNKNOWN"}
                  description="Network"
                  iconClass={
                    networkHealth ? "fa-check-circle" : "fa-exclamation-circle"
                  }
                  className={`smd-pointer`}
                  mode={networkHealth ? "success" : "danger"}
                />
              </Link>
              <br />
              <Link to="/blocks">
                <NumberCard
                  number={lastBlockNumber}
                  description="Blocks"
                  iconClass="fa-cube"
                />
              </Link>
            </div>
          </div>

          <div className="col-sm-3">
            <ValidatorsCard />
          </div>
          <div className="col-sm-6">
            <TransactionList />
          </div>
        </div>
        <div className="row">
          <div className="col-sm-12">
            <hr />
          </div>
        </div>
        <div className="row">
          <div className="col-sm-9 text-left">
            <h3>Historical Stats</h3>
          </div>
        </div>
        <div className="row">
          <div className="col-sm-3">
            <BarGraph
              data={txCount}
              number={txCount[0]}
              label={"Transactions per Last 15 Blocks"}
              identifier={"TxCount"}
            />
          </div>
          <div className="col-sm-3">
            <PieChart data={txTypeData} />
          </div>
          <div className="col-sm-3">
            <BarGraph
              data={blockPropData}
              units="s"
              label={"Block Interval Last 15 Blocks"}
              identifier={"BlockProp"}
            />
          </div>
        </div>
      </div>
    );
  }
}

export function mapStateToProps(state) {
  return {
    node: state.node,
    dashboard: state.dashboard,
    appMetadata: state.appMetadata,
  };
}

const connected = connect(mapStateToProps, {
  hideLoading,
  endTour,
  subscribeRoom,
  unSubscribeRoom,
  changeHealthStatus,
})(Dashboard);

export default withRouter(connected);
