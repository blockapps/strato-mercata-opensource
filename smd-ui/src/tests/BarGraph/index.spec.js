import React from 'react';
import BarGraph from '../../components/BarGraph';
import { dashboard } from '../Dashboard/dashboardMock';

describe('Test BarGraph index', () => {

  test('should render component properly', () => {
    const props = {
      data: dashboard.blockDifficulty,
      label: "Difficulty",
      identifier: "Difficulty"
    };

    const wrapper = render(
      <BarGraph {...props} />
    );

    expect(wrapper).toMatchSnapshot();
  })

})