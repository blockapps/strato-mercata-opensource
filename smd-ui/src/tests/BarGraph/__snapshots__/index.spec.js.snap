// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test BarGraph index should render component properly 1`] = `
initialize {
  "0": Object {
    "attribs": Object {
      "class": "pt-card pt-dark",
    },
    "children": Array [
      Object {
        "attribs": Object {
          "class": "row",
        },
        "children": Array [
          Object {
            "attribs": Object {
              "class": "col-sm-12 text-center",
            },
            "children": Array [
              Object {
                "attribs": Object {},
                "children": Array [
                  Object {
                    "data": "Difficulty",
                    "next": null,
                    "parent": [Circular],
                    "prev": null,
                    "type": "text",
                  },
                ],
                "name": "h4",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": null,
                "parent": [Circular],
                "prev": null,
                "type": "tag",
                "x-attribsNamespace": Object {},
                "x-attribsPrefix": Object {},
              },
            ],
            "name": "div",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": null,
            "parent": [Circular],
            "prev": null,
            "type": "tag",
            "x-attribsNamespace": Object {
              "class": undefined,
            },
            "x-attribsPrefix": Object {
              "class": undefined,
            },
          },
        ],
        "name": "div",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": Object {
          "attribs": Object {
            "class": "row",
          },
          "children": Array [
            Object {
              "attribs": Object {
                "class": "col-sm-12",
              },
              "children": Array [
                Object {
                  "attribs": Object {
                    "id": "bgDifficulty",
                    "style": "height:120px;",
                  },
                  "children": Array [],
                  "name": "div",
                  "namespace": "http://www.w3.org/1999/xhtml",
                  "next": null,
                  "parent": [Circular],
                  "prev": null,
                  "type": "tag",
                  "x-attribsNamespace": Object {
                    "id": undefined,
                    "style": undefined,
                  },
                  "x-attribsPrefix": Object {
                    "id": undefined,
                    "style": undefined,
                  },
                },
              ],
              "name": "div",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": null,
              "parent": [Circular],
              "prev": null,
              "type": "tag",
              "x-attribsNamespace": Object {
                "class": undefined,
              },
              "x-attribsPrefix": Object {
                "class": undefined,
              },
            },
          ],
          "name": "div",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": Object {
            "attribs": Object {
              "class": "row",
            },
            "children": Array [
              Object {
                "attribs": Object {
                  "class": "col-sm-12 text-center",
                },
                "children": Array [
                  Object {
                    "attribs": Object {},
                    "children": Array [
                      Object {
                        "data": "Avg:  118784",
                        "next": null,
                        "parent": [Circular],
                        "prev": null,
                        "type": "text",
                      },
                    ],
                    "name": "h2",
                    "namespace": "http://www.w3.org/1999/xhtml",
                    "next": null,
                    "parent": [Circular],
                    "prev": null,
                    "type": "tag",
                    "x-attribsNamespace": Object {},
                    "x-attribsPrefix": Object {},
                  },
                ],
                "name": "div",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": null,
                "parent": [Circular],
                "prev": null,
                "type": "tag",
                "x-attribsNamespace": Object {
                  "class": undefined,
                },
                "x-attribsPrefix": Object {
                  "class": undefined,
                },
              },
            ],
            "name": "div",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": null,
            "parent": [Circular],
            "prev": [Circular],
            "type": "tag",
            "x-attribsNamespace": Object {
              "class": undefined,
            },
            "x-attribsPrefix": Object {
              "class": undefined,
            },
          },
          "parent": [Circular],
          "prev": [Circular],
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
          },
        },
        "parent": [Circular],
        "prev": null,
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
        },
      },
      Object {
        "attribs": Object {
          "class": "row",
        },
        "children": Array [
          Object {
            "attribs": Object {
              "class": "col-sm-12",
            },
            "children": Array [
              Object {
                "attribs": Object {
                  "id": "bgDifficulty",
                  "style": "height:120px;",
                },
                "children": Array [],
                "name": "div",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": null,
                "parent": [Circular],
                "prev": null,
                "type": "tag",
                "x-attribsNamespace": Object {
                  "id": undefined,
                  "style": undefined,
                },
                "x-attribsPrefix": Object {
                  "id": undefined,
                  "style": undefined,
                },
              },
            ],
            "name": "div",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": null,
            "parent": [Circular],
            "prev": null,
            "type": "tag",
            "x-attribsNamespace": Object {
              "class": undefined,
            },
            "x-attribsPrefix": Object {
              "class": undefined,
            },
          },
        ],
        "name": "div",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": Object {
          "attribs": Object {
            "class": "row",
          },
          "children": Array [
            Object {
              "attribs": Object {
                "class": "col-sm-12 text-center",
              },
              "children": Array [
                Object {
                  "attribs": Object {},
                  "children": Array [
                    Object {
                      "data": "Avg:  118784",
                      "next": null,
                      "parent": [Circular],
                      "prev": null,
                      "type": "text",
                    },
                  ],
                  "name": "h2",
                  "namespace": "http://www.w3.org/1999/xhtml",
                  "next": null,
                  "parent": [Circular],
                  "prev": null,
                  "type": "tag",
                  "x-attribsNamespace": Object {},
                  "x-attribsPrefix": Object {},
                },
              ],
              "name": "div",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": null,
              "parent": [Circular],
              "prev": null,
              "type": "tag",
              "x-attribsNamespace": Object {
                "class": undefined,
              },
              "x-attribsPrefix": Object {
                "class": undefined,
              },
            },
          ],
          "name": "div",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": null,
          "parent": [Circular],
          "prev": [Circular],
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
          },
        },
        "parent": [Circular],
        "prev": Object {
          "attribs": Object {
            "class": "row",
          },
          "children": Array [
            Object {
              "attribs": Object {
                "class": "col-sm-12 text-center",
              },
              "children": Array [
                Object {
                  "attribs": Object {},
                  "children": Array [
                    Object {
                      "data": "Difficulty",
                      "next": null,
                      "parent": [Circular],
                      "prev": null,
                      "type": "text",
                    },
                  ],
                  "name": "h4",
                  "namespace": "http://www.w3.org/1999/xhtml",
                  "next": null,
                  "parent": [Circular],
                  "prev": null,
                  "type": "tag",
                  "x-attribsNamespace": Object {},
                  "x-attribsPrefix": Object {},
                },
              ],
              "name": "div",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": null,
              "parent": [Circular],
              "prev": null,
              "type": "tag",
              "x-attribsNamespace": Object {
                "class": undefined,
              },
              "x-attribsPrefix": Object {
                "class": undefined,
              },
            },
          ],
          "name": "div",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": [Circular],
          "parent": [Circular],
          "prev": null,
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
          },
        },
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
        },
      },
      Object {
        "attribs": Object {
          "class": "row",
        },
        "children": Array [
          Object {
            "attribs": Object {
              "class": "col-sm-12 text-center",
            },
            "children": Array [
              Object {
                "attribs": Object {},
                "children": Array [
                  Object {
                    "data": "Avg:  118784",
                    "next": null,
                    "parent": [Circular],
                    "prev": null,
                    "type": "text",
                  },
                ],
                "name": "h2",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": null,
                "parent": [Circular],
                "prev": null,
                "type": "tag",
                "x-attribsNamespace": Object {},
                "x-attribsPrefix": Object {},
              },
            ],
            "name": "div",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": null,
            "parent": [Circular],
            "prev": null,
            "type": "tag",
            "x-attribsNamespace": Object {
              "class": undefined,
            },
            "x-attribsPrefix": Object {
              "class": undefined,
            },
          },
        ],
        "name": "div",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": null,
        "parent": [Circular],
        "prev": Object {
          "attribs": Object {
            "class": "row",
          },
          "children": Array [
            Object {
              "attribs": Object {
                "class": "col-sm-12",
              },
              "children": Array [
                Object {
                  "attribs": Object {
                    "id": "bgDifficulty",
                    "style": "height:120px;",
                  },
                  "children": Array [],
                  "name": "div",
                  "namespace": "http://www.w3.org/1999/xhtml",
                  "next": null,
                  "parent": [Circular],
                  "prev": null,
                  "type": "tag",
                  "x-attribsNamespace": Object {
                    "id": undefined,
                    "style": undefined,
                  },
                  "x-attribsPrefix": Object {
                    "id": undefined,
                    "style": undefined,
                  },
                },
              ],
              "name": "div",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": null,
              "parent": [Circular],
              "prev": null,
              "type": "tag",
              "x-attribsNamespace": Object {
                "class": undefined,
              },
              "x-attribsPrefix": Object {
                "class": undefined,
              },
            },
          ],
          "name": "div",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": [Circular],
          "parent": [Circular],
          "prev": Object {
            "attribs": Object {
              "class": "row",
            },
            "children": Array [
              Object {
                "attribs": Object {
                  "class": "col-sm-12 text-center",
                },
                "children": Array [
                  Object {
                    "attribs": Object {},
                    "children": Array [
                      Object {
                        "data": "Difficulty",
                        "next": null,
                        "parent": [Circular],
                        "prev": null,
                        "type": "text",
                      },
                    ],
                    "name": "h4",
                    "namespace": "http://www.w3.org/1999/xhtml",
                    "next": null,
                    "parent": [Circular],
                    "prev": null,
                    "type": "tag",
                    "x-attribsNamespace": Object {},
                    "x-attribsPrefix": Object {},
                  },
                ],
                "name": "div",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": null,
                "parent": [Circular],
                "prev": null,
                "type": "tag",
                "x-attribsNamespace": Object {
                  "class": undefined,
                },
                "x-attribsPrefix": Object {
                  "class": undefined,
                },
              },
            ],
            "name": "div",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": [Circular],
            "parent": [Circular],
            "prev": null,
            "type": "tag",
            "x-attribsNamespace": Object {
              "class": undefined,
            },
            "x-attribsPrefix": Object {
              "class": undefined,
            },
          },
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
          },
        },
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
        },
      },
    ],
    "name": "div",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": null,
    "prev": null,
    "root": Object {
      "attribs": Object {},
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "type": "tag",
    "x-attribsNamespace": Object {
      "class": undefined,
    },
    "x-attribsPrefix": Object {
      "class": undefined,
    },
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;
